### ISSUE TITLE
Unable to validate my configuration using the curve25519-sha256 from SFTPKeyExchanges

### ISSUE THREAD

**Issue by frenchieisverige** [proftpd/proftpd#1635]
Hi all 👋

### What I Did

I tried to validate my configuration with `proftpd -t` but I always have the following error :
```bash
 fatal: SFTPKeyExchanges: unsupported key exchange algorithm: curve25519-sha256 on line xx of '/etc/proftpd/conf.d/sftp.conf'
```

### What I Expected/Wanted

I would like to perform a key exchange using the `curve25519-sha256` algorithm and some explainations on what I'm doing wrong 😅

### ProFTPD Version and Configuration

Here is the environment I use to set up this :

- a fresh CT of Ubuntu 22.04 (Proxmox)
- `proftpd` and `proftpd-mod-crypto` are installed
- oob `proftpd.conf` in `/etc/proftpd`
- `modules.conf` : 
```
# Install proftpd-mod-crypto to use this module for SFTP support.
LoadModule mod_sftp.c
#LoadModule mod_sftp_pam.c
```
- and last but not least my `sftp.conf` located in the subdirectory `conf.d` :
```
<IfModule mod_sftp.c>

  Port 2299
  UseSendfile on
  DefaultRoot ~

  SFTPEngine on
  SFTPLog /var/log/proftpd/sftp.log
  SFTPAuthMethods publickey password
  SFTPTrafficPolicy low
  SFTPHostKey /etc/ssh/ssh_host_ecdsa_key
  SFTPHostKey /etc/ssh/ssh_host_ed25519_key
  SFTPHostKey /etc/ssh/ssh_host_rsa_key
  AuthOrder mod_auth_file.c
  AuthUserFile /etc/proftpd/ftpd.passwd
  AuthGroupFile /etc/proftpd/ftpd.group
  SFTPAuthorizedUserKeys file:/etc/proftpd/authorized_keys/%u
  SFTPCompression delayed
  SFTPKeyExchanges curve25519-sha256 diffie-hellman-group-exchange-sha256 ecdh-sha2-nistp521 ecdh-sha2-nistp384 ecdh-sha2-nistp256
  SFTPCiphers aes256-gcm@openssh.com aes256-ctr aes192-ctr aes128-ctr aes256-cbc
  SFTPDigests hmac-sha2-256-etm@openssh.com hmac-sha2-256 hmac-sha2-512 hmac-ripemd160 hmac-md5
</IfModule>

<IfModule mod_sftp_pam.c>
  SFTPPAMEngine off
</IfModule>
```

Moreover, here is the result of proftpd -V
```
Compile-time Settings:
  Version: 1.3.7c (maint)
  Platform: LINUX [Linux 5.15.74-1-pve x86_64]
  Built: Fri Dec 3 2021 13:35:22 UTC
  Built With:
    configure  '--infodir=/share/info' '--disable-option-checking' '--disable-silent-rules' '--libdir=/lib/x86_64-linux-gnu' '--disable-dependency-tracking' '--prefix=/usr' '--with-pkgconfig=lib/pkgconfig' '--with-includes=/usr/include/postgresql:/usr/include/mysql' '--mandir=/usr/share/man' '--sysconfdir=/etc/proftpd' '--localstatedir=/run' '--libexecdir=/usr/lib/proftpd' '--enable-sendfile' '--enable-facl' '--enable-dso' '--enable-autoshadow' '--enable-ctrls' '--enable-openssl' '--enable-ipv6' '--enable-nls' '--enable-memcache' '--with-lastlog=/var/log/lastlog' '--enable-pcre' '--disable-strip' '--enable-redis' '--build' 'x86_64-linux-gnu' '--with-shared=mod_unique_id:mod_site_misc:mod_load:mod_ban:mod_quotatab:mod_sql:mod_sql_mysql:mod_sql_postgres:mod_sql_sqlite:mod_sql_odbc:mod_dynmasq:mod_quotatab_sql:mod_ldap:mod_quotatab_ldap:mod_ratio:mod_tls:mod_rewrite:mod_radius:mod_wrap:mod_wrap2:mod_wrap2_file:mod_wrap2_sql:mod_quotatab_file:mod_quotatab_radius:mod_facl:mod_ctrls_admin:mod_copy:mod_deflate:mod_ifversion:mod_geoip:mod_exec:mod_sftp:mod_sftp_pam:mod_sftp_sql:mod_shaper:mod_sql_passwd:mod_ifsession:mod_auth_otp:mod_tls_redis:mod_wrap2_redis:mod_redis:mod_memcache:mod_tls_memcache:mod_readme:mod_snmp:mod_digest:mod_ident:mod_log_forensic:mod_qos:mod_statcache:mod_tls_fscache:mod_tls_shmcache:mod_dnsbl' 'build_alias=x86_64-linux-gnu' 'CFLAGS=-g -O2 -ffile-prefix-map=/build/proftpd-dfsg-iPbrOs/proftpd-dfsg-1.3.7c+dfsg=. -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security' 'LDFLAGS=-Wl,-Bsymbolic-functions -flto=auto -Wl,-z,relro -Wl,-z,now' 'CPPFLAGS=-Wdate-time -D_FORTIFY_SOURCE=2' 'CXXFLAGS=-g -O2 -ffile-prefix-map=/build/proftpd-dfsg-iPbrOs/proftpd-dfsg-1.3.7c+dfsg=. -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security'

  CFLAGS: -g2 -g -O2 -ffile-prefix-map=/build/proftpd-dfsg-iPbrOs/proftpd-dfsg-1.3.7c+dfsg=. -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security -Wall -fno-omit-frame-pointer -fno-strict-aliasing -Werror=implicit-function-declaration
  LDFLAGS: -L$(top_srcdir)/lib -L$(top_builddir)/lib -Wl,-Bsymbolic-functions -flto=auto -Wl,-z,relro -Wl,-z,now -rdynamic  -L/usr/lib/x86_64-linux-gnu -L/usr/lib/x86_64-linux-gnu
  LIBS: -lacl  -lpcreposix -lpcre -lssl -lcrypto -lsodium -lcap  -lpam -lsupp -lattr -lnsl -lresolv -lresolv -lcrypt -lhiredis -lmemcachedutil -lmemcached  -pthread

  Files:
    Configuration File:
      /etc/proftpd/proftpd.conf
    Pid File:
      /run/proftpd.pid
    Scoreboard File:
      /run/proftpd.scoreboard
    Header Directory:
      /usr/include/proftpd
    Shared Module Directory:
      /usr/lib/proftpd

  Info:
    + Max supported UID: 4294967295
    + Max supported GID: 4294967295

  Features:
    + Autoshadow support
    + Controls support
    + curses support
    - Developer support
    + DSO support
    + IPv6 support
    + Largefile support
    + Lastlog support
    + Memcache support
    + ncursesw support
    + NLS support
    + OpenSSL support (OpenSSL 3.0.0 7 sep 2021)
    + PCRE support
    + POSIX ACL support
    + Redis support
    + Sendfile support
    + Shadow file support
    + Sodium support
    + Trace support
    + xattr support

  Tunable Options:
    PR_TUNABLE_BUFFER_SIZE = 1024
    PR_TUNABLE_DEFAULT_RCVBUFSZ = 8192
    PR_TUNABLE_DEFAULT_SNDBUFSZ = 8192
    PR_TUNABLE_ENV_MAX = 2048
    PR_TUNABLE_GLOBBING_MAX_MATCHES = 100000
    PR_TUNABLE_GLOBBING_MAX_RECURSION = 8
    PR_TUNABLE_HASH_TABLE_SIZE = 40
    PR_TUNABLE_LOGIN_MAX = 256
    PR_TUNABLE_NEW_POOL_SIZE = 512
    PR_TUNABLE_PATH_MAX = 4096
    PR_TUNABLE_SCOREBOARD_BUFFER_SIZE = 80
    PR_TUNABLE_SCOREBOARD_SCRUB_TIMER = 30
    PR_TUNABLE_SELECT_TIMEOUT = 30
    PR_TUNABLE_TIMEOUTIDENT = 10
    PR_TUNABLE_TIMEOUTIDLE = 600
    PR_TUNABLE_TIMEOUTLINGER = 10
    PR_TUNABLE_TIMEOUTLOGIN = 300
    PR_TUNABLE_TIMEOUTNOXFER = 300
    PR_TUNABLE_TIMEOUTSTALLED = 3600
    PR_TUNABLE_XFER_SCOREBOARD_UPDATES = 10
```

Thanks in advance !

**Comment by Castaglia** [comment]
Interesting.  What do you mean, exactly, by "a fresh CT of Ubuntu 22.04 (Proxmox)"?  Is there a Docker image that I might reference, to reproduce this behavior locally?

**Comment by Castaglia** [comment]
OK, I'm able to reproduce this locally, using a stock `ubuntu:22.04` Docker image.

In the 1.3.7x ProFTPD series, the key exchange name "curve25519-sha256" is *not* supported.  Early drafts of the RFCs for this only used the form "curve25519-sha256@libssh.org", because it was a proposed extension.  In ProFTPD 1.3.8 and later, the name "curve25519-sha256" is also supported -- and both refer to the same functionality

**Comment by frenchieisverige** [comment]
I was about to say that I can reproduce it also on Docker, but you were faster 😅

I was biased by the documentation that says ProFTPd supports it from 1.3.2rc2 and later... 

So if I understand you correctly, I can still use the curve25519-sha256 algorithm with [curve25519-sha256@libssh.org](mailto:curve25519-sha256@libssh.org) ?

PS : CT from Proxmox is also a containerization technology running LXC. If you want to read more about it : https://pve.proxmox.com/wiki/Linux_Container

**Comment by Castaglia** [comment]
> So if I understand you correctly, I can still use the curve25519-sha256 algorithm with `curve25519-sha256@libssh.org`?

That's correct.

**Comment by frenchieisverige** [comment]
Thanks. 
One more question though : What about my _SFTPCiphers_ and _SFTPDigests_  directives ? Same story ?

**Comment by Castaglia** [comment]
In your `SFTPCiphers` configuration, the "aes256-gcm@openssh.com" algorithm support was implemented in ProFTPD 1.3.8rc3 (see [Bug #3759](http://bugs.proftpd.org/show_bug.cgi?id=3759)); you will thus need to remove that from your list.

**Comment by Castaglia** [comment]
If it helps, what I used to determine this was:
```
$ proftpd -td10
```
The `-t` command-line option says to do to a syntax check ("test") of the configuration (including `Include` files).  The `-d10` says to emit debug level messages (level 10) while doing so.   It's what I often use to ensure that a given configuration passes initial sanity checks, before using that configuration at runtime.

**Comment by frenchieisverige** [comment]
> In your `SFTPCiphers` configuration, the "[aes256-gcm@openssh.com](mailto:aes256-gcm@openssh.com)" algorithm support was implemented in ProFTPD 1.3.8rc3 (see [Bug #3759](http://bugs.proftpd.org/show_bug.cgi?id=3759)); you will thus need to remove that from your list.
> 
> On my local Docker container, your `SFTPDigests` configuration passes without warning/error.
Hmm, weird. On my side, ProFTPd is still complaning about `hmac-sha2-256-etm@openssh.com` 🤔
```
fatal: SFTPDigests: unsupported digest algorithm: hmac-sha2-256-etm@openssh.com on line xx of '/etc/proftpd/conf.d/sftp.conf'
```




> If it helps, what I used to determine this was:
> 
> ```
> $ proftpd -td10
> ```
> 
> The `-t` command-line option says to do to a syntax check ("test") of the configuration (including `Include` files). The `-d10` says to emit debug level messages (level 10) while doing so. It's what I often use to ensure that a given configuration passes initial sanity checks, before using that configuration at runtime.
Thanks for the tip. Highly appriciated!

**Comment by Castaglia** [comment]
> fatal: SFTPDigests: unsupported digest algorithm: hmac-sha2-256-etm@openssh.com on line xx of '/etc/proftpd/conf.d/sftp.conf'

Ah, right.   I see this as well.  The ETM (Encrypt-Then-Mac) extensions were implemented in ProFTPD 1.3.8rc3; see Issue #1330.  So you may need to remove that as well from your `SFTPDigests` list.

**Comment by frenchieisverige** [comment]
Alright. At the end of the day, I have to wait for version 1.3.8 in Ubuntu's repos, or compile it myself.
Thank you again for your help! 👍
